sdk/storage/Azure.Storage.DataMovement/src/TransferItemScheduler.cs

```diff
@@ -0,0 +1,156 @@
+            }
+
+            // Set _rangeSize
+            if (transferOptions.MaximumTransferSize.HasValue
+                && transferOptions.MaximumTransferSize.Value > 0)
+            {
+                _rangeSize = Math.Min(transferOptions.MaximumTransferSize.Value, Constants.Blob.Block.MaxDownloadBytes);
+            }
+            else
+            {
+                _rangeSize = Constants.DefaultBufferSize;
+            }
+
+            // Set _initialRangeSize
+            if (transferOptions.InitialTransferSize.HasValue
+                && transferOptions.InitialTransferSize.Value > 0)
+            {
+                _initialDownloadRangeSize = transferOptions.InitialTransferSize.Value;
+            }
+            else
+            {
+                _initialDownloadRangeSize = Constants.Blob.Block.DefaultInitalDownloadRangeSize;
+            }
+
+            // Set _singleUploadThreshold
+            if (transferOptions.InitialTransferSize.HasValue
+                && transferOptions.InitialTransferSize.Value > 0)
+            {
+                _singleUploadThreshold = Math.Min(transferOptions.InitialTransferSize.Value, Constants.Blob.Block.MaxUploadBytes);
+            }
+            else
+            {
+                _singleUploadThreshold = Constants.Blob.Block.Pre_2019_12_12_MaxUploadBytes;
+            }
+        }
+
+        /// <summary>
+        /// Adds job to the queue that has already been scanned.
+        /// </summary>
+        /// <param name="job"></param>
+        public void AddJob(StorageTransferJob job)
+        {
+            itemsToProcess.Enqueue(job);
+        }
+
+        /// <summary>
+        /// Runs jobs on the multiple threads given.
+        ///
+        /// This will be called once there's one job on the queue, if there are no items in the queue then we exit.
+        /// </summary>
+        private async Task RunJobsAsync()
+        {
+            // Create a list of tasks that will each run each Transfer Item
+            // of a Job.  The queue maintains the order of the Tasks
+            // so we can keep appending to the end of the destination
+            // stream when each segment finishes.
+            var runningTasks = new List<Task>();
+            while (!itemsToProcess.IsEmpty)
+            {
+                StorageTransferJob currentJob = await itemsToProcess.DequeueAsync().ConfigureAwait(false);
+
+                // Create Task from given job type
+                // TODO: remove this and call the according job type to create the task
+                // Add Task
+                runningTasks.Add(currentJob.CreateTransferTaskAsync());
+
+                // If we run out of workers
+                if (runningTasks.Count >= _maxWorkerCount)
+                {
+                    // Wait for at least one of them to finish
+                    await Task.WhenAny(runningTasks).ConfigureAwait(false);
```

Comment:
As it is, looks like concurrency customizations only allow for modifying the number of jobs running in parallel, but not the number of items within a job running in parallel. Is that something to consider/is already on the plans? A customer might prioritize an entire job being finished before beginning another job, rather than pushing them all to completion. In that case, concurrency wouldn't benefit them right now, since they wouldn't be able to, say, set a `maxJobConcurrency` to 1 and a `maxOperationConcurrency` to 5, which could let DMLib process one job at a time and run through uploading/downloading 5 items in the job at one time.

That wouldn't need any changes here, just adding this same sort of thread handling inside a job's `CreateTransferTaskAsync`/`StartTransferAsync` working off of a `maxOperationConcurrency` option. Those two options could just sit in the StorageTransferOptions. Maybe have jobs get the options from the scheduler they're in when started to prevent conflicts in settings too?

Reply:
You're right that's what it looks like right now, but I should have added a comment here that we would eventually move to adding all the Tasks / Transfer Items from a Transfer Job to run in parallel. That way it's not running jobs in parallel, its running each transfer item in parallel.

That being said, this is also a balancing act because we also have to make sure we allocate for each transfer item and how many threads that take since we would be most likely taking advantage of the parallel upload and/or download scheme. I will add in a comment about this as well.